File: ports/go/core/contextwindow.go
package core

import "sort"

// Priority is the MoSCoW priority of a context item.
type Priority int

const (
	PriorityWont   Priority = 0
	PriorityCould  Priority = 1
	PriorityShould Priority = 2
	PriorityMust   Priority = 3
)

// ContextItem is a prioritized piece of context with a cheap token estimate.
type ContextItem struct {
	Text     string
	Priority Priority
}

// Tokens is a deterministic ~4-chars/token estimate.
func (c ContextItem) Tokens() int {
	t := (len(c.Text) + 3) / 4
	if t < 1 {
		return 1
	}
	return t
}

// ContextWindowManager compacts items to a token budget, MoSCoW-first.
type ContextWindowManager struct{ MaxTokens int }

// NewContextWindowManager builds a manager (maxTokens<=0 => 1).
func NewContextWindowManager(maxTokens int) *ContextWindowManager {
	if maxTokens <= 0 {
		maxTokens = 1
	}
	return &ContextWindowManager{MaxTokens: maxTokens}
}

// Compact drops WON'T items, then greedily keeps highest-priority items within budget,
// preserving the original order among kept items.
func (m *ContextWindowManager) Compact(items []ContextItem) []ContextItem {
	var cands []ContextItem
	for _, it := range items {
		if it.Priority != PriorityWont {
			cands = append(cands, it)
		}
	}
	order := make([]int, len(cands))
	for i := range order {
		order[i] = i
	}
	sort.SliceStable(order, func(a, b int) bool {
		return cands[order[a]].Priority > cands[order[b]].Priority
	})
	budget := m.MaxTokens
	keep := make(map[int]bool)
	for _, i := range order {
		if cands[i].Tokens() <= budget {
			keep[i] = true
			budget -= cands[i].Tokens()
		}
	}
	var out []ContextItem
	for i := range cands {
		if keep[i] {
			out = append(out, cands[i])
		}
	}
	return out
}

// TotalTokens sums the token estimates.
func (m *ContextWindowManager) TotalTokens(items []ContextItem) int {
	total := 0
	for _, it := range items {
		total += it.Tokens()
	}
	return total
}

File: agentic-clj/src/agentic/context_window.clj
(ns agentic.context-window
  "Context-window management — MoSCoW compaction so a transcript fits a token budget. Keep MUST
   first, then SHOULD, then COULD; drop WONT; stop at the budget; preserve input order among kept
   items. Mirror of jagentic-core ContextWindowManager. Items are {:text :priority} with priority
   in #{:wont :could :should :must}.")

(def ^:private priority-rank {:wont 0 :could 1 :should 2 :must 3})

(defn tokens
  "Cheap ~4-chars/token estimate."
  [text]
  (max 1 (quot (+ (count text) 3) 4)))

(defn compact
  "Return the kept items (original order) whose total token estimate fits max-tokens, dropping
   :wont and preferring higher priority."
  [items max-tokens]
  (let [budget (max 1 max-tokens)
        candidates (vec (keep-indexed (fn [i it] (when (not= :wont (:priority it)) [i it]))
                                      items))
        ;; choose which to keep: descending priority, stable on original index
        order (sort-by (fn [[i it]] [(- (priority-rank (:priority it) 0)) i]) candidates)
        kept (loop [left budget [pair & more] order keep #{}]
               (if (nil? pair)
                 keep
                 (let [[i it] pair t (tokens (:text it))]
                   (if (<= t left)
                     (recur (- left t) more (conj keep i))
                     (recur left more keep)))))]
    (->> candidates
         (filter (fn [[i _]] (kept i)))
         (map second)
         vec)))
